src/proper_gen.erl

```diff
@@ -406,12 +406,46 @@ float_gen(_Size, Low, High) ->
 %% We make sure we never clash with internal atoms by checking that the first
 %% character is not '$'.
 atom_gen(Size) ->
+    atom_gen(get('$atom_limit'), Size).
+
+atom_gen(infinity, Size) ->
+    atom_gen1(Size);
+atom_gen(Limit, Size) ->
+    ?LAZY(
+        ?LET(Atom,
+             try persistent_term:get({'$proper', ?MODULE, ?FUNCTION_NAME, Size}) of
+                 {N, Atoms} when N =< Limit ->
+                     proper_types:oneof(Atoms);
+                 {N, CurAtoms} ->
+                     ?LET(MoreAtoms,
+                          proper_types:vector(Limit-N, atom_gen1(Size)),
+                          begin
+                              Atoms = CurAtoms ++ MoreAtoms,
+                              persisten_term:put({'$proper', ?MODULE, ?FUNCTION_NAME, Size},
```

Comment:
This line cannot possibly have ever executed...

Reply:
Yeah, just noticed myself :sweat_smile:

I probably should have said that this is not yet finished, just wanted to show it to see if the general idea will be accepted. Therefore, there are no tests yet.

Reply:
Typo corrected.